### ISSUE TITLE
Option + impl Trait = uncompilable code (?)

### ISSUE THREAD

**Issue by japaric** [rust-lang/rust#37454]
### STR

``` rust
#![feature(conservative_impl_trait)]

trait Foo {
    fn foo(&self);
}

struct FooStruct {}

impl Foo for FooStruct {
    fn foo(&self) {
        println!("foo");
    }
}

fn mk_foo() -> impl Foo {
    FooStruct {}
}

fn main() {
    let mut maybe_foo = None;

    loop {
        if some_condition() {
            maybe_foo.take().map(|f| f.foo());

            maybe_foo = Some(mk_foo());
        }

        // .. more code ..
    }
}

fn some_condition() -> bool {
    true
}
```

```
rustc foo.rs
error: the type of this value must be known in this context
  --> foo.rs:24:38
   |
24 |             maybe_foo.take().map(|f| f.foo());
   |                                      ^^^^^^^

error: aborting due to previous error
```

---

This works:

``` diff
     loop {
         if true {
-            maybe_foo.take().map(|f| f.foo());
-
             maybe_foo = Some(mk_foo());
+
+            maybe_foo.take().map(|f| f.foo());
         }
```

but changes the behaviour of the code.

My first attempt to fix this was:

``` diff
 fn main() {
-    let mut maybe_foo = None;
+    let mut maybe_foo: Option<impl Foo> = None;

     loop {
```

but: 

```
error[E0562]: `impl Trait` not allowed outside of function and inherent method return types
```

My second attempt was:

``` diff

+fn init_maybe_foo() -> Option<impl Foo> {
+    None::<FooStruct>
+}
+
 fn main() {
-    let mut maybe_foo = None;
+    let mut maybe_foo = init_maybe_foo();

     loop {
```

but:

```
$ rustc foo.rs
error[E0308]: mismatched types
  --> foo.rs:30:30
   |
30 |             maybe_foo = Some(mk_foo());
   |                              ^^^^^^^^ expected anonymized type, found a different anonymized type
   |
   = note: expected type `impl Foo` (anonymized type)
   = note:    found type `impl Foo` (anonymized type)

error: aborting due to previous error
```

What? They have the same type.

Are `impl Trait` types in `let` statements planned? Or can type inference be ... improved (?) to handle the last case? Or does anyone know how to work around this? :smile: 

cc @eddyb

**Comment by eddyb** [comment]
```rust
fn hint_none<R, F: FnOnce() -> R>(_: F) -> Option<T> {None}
...
let mut maybe_foo = hint_none(mk_foo);
```

**Comment by mister-walter** [comment]
Is this blocked on the implemention of rust-lang/rfcs#2071 ?

**Comment by cramertj** [comment]
This isn't a full solution,  but the code in the original issue report can be fixed using UFCS: `Foo::foo(&f)` works.

**Comment by Dylan-DPC** [comment]
Current output with code adjusted [playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=fe0a6952f8c6b09815f4fcefa78fbd15) 

```
error[E0282]: type annotations needed
  --> src/main.rs:22:35
   |
22 |             maybe_foo.take().map(|f| f.foo());
   |                                   ^  - type must be known at this point
   |
help: consider giving this closure parameter an explicit type
   |
22 |             maybe_foo.take().map(|f: /* Type */| f.foo());
   |                                    ++++++++++++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

**Comment by Dylan-DPC** [comment]
Closing this as the error has changed since the issue was created. I feel even if this issue is not resolved in any way, a new issue with fresh code and expectations might be a  better idea than a long forgotten buried issue. Thanks
